### PR TITLE
[C++ Interop] Handle importing nested types and namespaces better.

### PR DESCRIPTION
**Pull request by pschuh**
Note that namespaces should require mangling support in the future for
the generated swift wrapping type.

### CODE REVIEW

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -1485,10 +1485,15 @@ ASTMangler::getSpecialManglingContext(const ValueDecl *decl,
       else
         hasNameForLinkage = !clangDecl->getDeclName().isEmpty();
       if (hasNameForLinkage) {
-        auto *clangDC = clangDecl->getDeclContext();
-        assert(clangDC->getRedeclContext()->isTranslationUnit() &&
-               "non-top-level Clang types not supported yet");
-        (void)clangDC;
+        if (!decl->getASTContext().LangOpts.EnableCXXInterop) {
+          // TODO: Mangle namespaces into these ObjC names.
```
Comment (bgogul) [inline lib/AST/ASTMangler.cpp]: Add a bug for the TODO?

File: lib/ClangImporter/ImportDecl.cpp [hunk 2]
```diff
@@ -8066,7 +8072,7 @@ ClangImporter::Implementation::importDeclContextOf(
   switch (context.getKind()) {
   case EffectiveClangContext::DeclContext: {
     auto dc = context.getAsDeclContext();
-    if (dc->isTranslationUnit()) {
+    if (dc->isFileContext()) {
```
Comment (bgogul) [inline lib/ClangImporter/ImportDecl.cpp]: Why is this change needed?

File: lib/ClangImporter/ImportDecl.cpp [hunk 3]
```diff
@@ -3162,6 +3162,12 @@ namespace {
 
       // Create the struct declaration and record it.
       auto name = importedName.getDeclName().getBaseIdentifier();
+      {
+        // Hack for nested types (They produce cycles)...
```
Comment (bgogul) [inline lib/ClangImporter/ImportDecl.cpp]: Can you add a small example here?

**Review by bgogul — COMMENTED** [review]
Had a couple of clarifications.

File: lib/ClangImporter/ImportDecl.cpp (hunk 2 above)
Comment (pschuh) [reply lib/ClangImporter/ImportDecl.cpp]: This is needed to support namespaces. Decls inside a namespace are treated as toplevel. This is not used for purposes of lookup (you have to alias the name into the toplevel scope in order for it to be findable).

File: test/ClangImporter/Inputs/custom-modules/cxx_interop.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+ns::Basic makeA();
+
+using ns_Basic = ns::Basic;
+
+namespace nested_struct_problem {
+
+class WrapperClass {
+ public:
+  struct NestedStruct {
```
Comment (bgogul) [inline test/ClangImporter/Inputs/custom-modules/cxx_interop.h]: Can you add small comment as to why this is a problem? Say "Referring to NestedStruct before WrapperClass can trigger an infinite loop in ImportDecl" or something along the lines?

**Review by bgogul — APPROVED** [review]

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by dan-zheng** [comment]
`test/ClangImporter/cxx_interop.swift` fails on macOS:
```
FAIL: Swift(macosx-x86_64) :: ClangImporter/cxx_interop.swift (1 of 1)
******************** TEST 'Swift(macosx-x86_64) :: ClangImporter/cxx_interop.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/ClangImporter/Output/cxx_interop.swift.tmp" && mkdir -p "/Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/ClangImporter/Output/cxx_interop.swift.tmp"
: 'RUN: at line 2';   /Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/bin/swiftc -frontend -target x86_64-apple-macosx10.9  -module-cache-path '/Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' -sdk '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk' -swift-version 4   -typecheck /Users/danielzheng/swift-dev/swift/test/ClangImporter/cxx_interop.swift -I /Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules -module-cache-path /Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/ClangImporter/Output/cxx_interop.swift.tmp -enable-cxx-interop
--
Exit Code: 1

Command Output (stderr):
--
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:3:1: error: unknown type name 'namespace'
namespace ns {
^
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:3:13: error: expected ';' after top level declarator
namespace ns {
            ^
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:13:1: error: unknown type name 'using'
using ns_Basic = ns::Basic;
^
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:15:1: error: unknown type name 'namespace'
namespace nested_struct_problem {
^
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:15:32: error: expected ';' after top level declarator
namespace nested_struct_problem {
                               ^
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "cxx_interop.h"
        ^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/Inputs/custom-modules/cxx_interop.h:30:1: error: unknown type name 'NestedStruct'
NestedStruct MakeNestedStruct();
^
/Users/danielzheng/swift-dev/swift/test/ClangImporter/cxx_interop.swift:4:8: error: could not build Objective-C module 'CXXInterop'
import CXXInterop
       ^

--

********************
Testing Time: 1.06s
********************
Failing Tests (1):
    Swift(macosx-x86_64) :: ClangImporter/cxx_interop.swift
```

cc @pschuh: do you have any ideas about why this fails on macOS?
